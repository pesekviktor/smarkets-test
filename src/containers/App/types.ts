import {ActionType} from 'typesafe-actions';
import * as actions from './actions';

/* --- STATE --- */

export interface AppState {
  readonly eventsState: EventsState;
}

export interface EventsPagination {
  /**
   * Pagination string tells where to go for next page
   */
  next_page: string;
}

export interface EventsResponse {
  events: SmarketsEvent[];
  pagination?: EventsPagination;
}

export interface SmarketsEvent {
  id: string;
  name: string;
  short_name?: string;
  start_datetime: string;
  description: string;
}

export interface EventsState {
  /**
   * In more complex project setup I prefer to use
   * an enumerated state (by TS) (INIT, FETCHING, ERROR) which can be extended
   * rather then loading.
   */
  loading: boolean;
  readonly error?: object | boolean;
  results: SmarketsEvent[];
  pagination?: EventsPagination;
}

export interface EventFilter {
  type?: string;
  paginationUrl?: string;
}

/* --- ACTIONS --- */
export type AppActions = ActionType<typeof actions>;
