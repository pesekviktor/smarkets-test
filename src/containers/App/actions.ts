import {action} from 'typesafe-actions';

import ActionTypes from './constants';
import {EventFilter, EventsResponse} from './types';

export const fetchEvents = (eventFilter: EventFilter) =>
  action(ActionTypes.FETCH_EVENTS, eventFilter);
export const eventsFetched = (response: EventsResponse) =>
  action(ActionTypes.FETCH_EVENTS_SUCCESS, {response: response});
export const eventsFetchError = (error: object) =>
  action(ActionTypes.FETCH_EVENTS_ERROR, error);
