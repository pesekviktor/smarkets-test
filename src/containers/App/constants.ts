enum ActionTypes {
  FETCH_EVENTS = 'smarkets/App/FETCH_EVENTS',
  FETCH_EVENTS_SUCCESS = 'smarkets/App/FETCH_EVENTS_SUCCESS',
  FETCH_EVENTS_ERROR = 'smarkets/App/FETCH_EVENTS_ERROR',
}

export default ActionTypes;
