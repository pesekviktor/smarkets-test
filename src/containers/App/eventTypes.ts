export const EVENT_TYPES = [
  'american_football_match',
  'american_football_outright',
  'baseball_match',
  'baseball_outright',
  'basketball_match',
  'boxing_match',
  'cricket_match',
  'cricket_outright',
  'current_affairs',
  'cycling',
  'darts_match',
  'darts_outright',
  'football_match',
  'football_outright',
  'golf_match',
  'golf_outright',
  'handball_match',
  'horse_racing_race',
  'ice_hockey_match',
  'mma_match',
  'motorsports_race',
  'motorsports_outright',
  'politics',
  'politics_outright',
  'rowing',
  'rugby_league_match',
  'rugby_league_outright',
  'rugby_union_match',
  'rugby_union_outright',
  'snooker_match',
  'snooker_outright',
  'tennis_match',
  'tennis_outright',
  'volleyball_match',
  'generic',
  'top_level_event',
  'tv_entertainment',
];
