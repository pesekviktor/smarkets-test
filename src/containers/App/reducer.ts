import {AppActions, AppState} from './types';
import ActionTypes from './constants';

export const initialState: AppState = {
  eventsState: {
    results: [],
    error: false,
    loading: false,
  },
};

function appReducer(
  state: AppState = initialState,
  action: AppActions,
): AppState {
  switch (action.type) {
    case ActionTypes.FETCH_EVENTS:
      return {
        eventsState: {
          loading: true,
          error: false,
          results: [],
        },
      };
    case ActionTypes.FETCH_EVENTS_SUCCESS:
      return {
        eventsState: {
          loading: false,
          error: state.eventsState.error,
          results: action.payload.response.events,
          pagination: action.payload.response.pagination,
        },
      };
    case ActionTypes.FETCH_EVENTS_ERROR:
      const {eventsState, ...rest} = state;
      return {
        eventsState: {
          ...eventsState,
          loading: false,
          error: action.payload,
        },
        ...rest,
      };
    default:
      return state;
  }
}

export default appReducer;
