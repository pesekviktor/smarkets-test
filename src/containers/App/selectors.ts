import {createSelector} from 'reselect';
import {ApplicationRootState} from '../../types';
import {AppState} from "./types";

const selectAppState = (state: ApplicationRootState) => state.appState;

export const makeSelectEventsLoading = () =>
  createSelector(selectAppState, appState => appState.eventsState.loading);

export const makeSelectEventsError = () =>
  createSelector(selectAppState, appState => appState.eventsState.error);

export const makeSelectEvents = () =>
  createSelector(selectAppState, (appState: AppState) => appState.eventsState.results);

export const makeSelectEventsPagination = () =>
  createSelector(selectAppState, appState => appState.eventsState.pagination);

export const makeSelectEventsState = () =>
  createSelector(selectAppState, appState => appState.eventsState);
