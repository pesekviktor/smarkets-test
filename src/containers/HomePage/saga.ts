import {call, put, takeLatest} from 'redux-saga/effects';
import {eventsFetched, eventsFetchError, fetchEvents,} from '../App/actions';
import ActionTypes from '../App/constants';

import request from '../../utils/request';
import {ActionType} from 'typesafe-actions';

export function* getUpcomingEvents({
                                     payload: {paginationUrl, type},
                                   }: ActionType<typeof fetchEvents>) {
  /**
   * In complex application urls would be in environment variables and paths would be kept in some api constants.
   * There would be a library to build url params etc.
   */
  let requestURL = `/v3/events/?state=upcoming&type=${type}`;
  if (paginationUrl) {
    requestURL = `/v3/events/${paginationUrl}`;
  }

  try {
    const events = yield call(request, requestURL);
    yield put(eventsFetched(events));
  } catch (err) {
    yield put(eventsFetchError(err));
  }
}

export default function* fetchUpcomingEvents() {
  yield takeLatest(ActionTypes.FETCH_EVENTS, getUpcomingEvents);
}
