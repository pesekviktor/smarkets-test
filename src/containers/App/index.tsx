/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import * as React from 'react';
import {Route, Switch} from 'react-router-dom';
import {hot} from 'react-hot-loader/root';

import HomePage from '../HomePage/Loadable';
import NotFoundPage from '../NotFoundPage/Loadable';

import GlobalStyle from '../../global-styles';
import styled from 'styled-components';
import {Helmet} from 'react-helmet-async';

const AppWrapper = styled.div``;

function App() {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - Smarkets events"
        defaultTitle="Smarkets events"
      >
        <meta
          name="description"
          content="An applications showing smarkets events"
        />
      </Helmet>
      <Switch>
        <Route exact path="/" component={HomePage}/>
        <Route component={NotFoundPage}/>
      </Switch>
      <GlobalStyle/>
    </AppWrapper>
  );
}

export default hot(App);
