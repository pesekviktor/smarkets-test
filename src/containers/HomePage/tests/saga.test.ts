/**
 * Tests for HomePage sagas
 */

import {put} from 'redux-saga/effects';

import ActionTypes from '../../App/constants';

import {getUpcomingEvents} from '../saga';
import {EventsResponse} from "../../App/types";
import {eventsFetched, eventsFetchError} from "../../App/actions";


describe('getEvents Saga', () => {
  let getUpcomingEventsGenerator;

  beforeEach(() => {
    getUpcomingEventsGenerator = getUpcomingEvents({
      type: ActionTypes.FETCH_EVENTS, payload: {
        type: 'football'
      }
    });

    const callDescriptor = getUpcomingEventsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the eventsFetched action if it requests the data successfully', () => {
    const response = {
      pagination: {next_page: 'next_page',},
      events: [{short_name: 'name'}]
    } as EventsResponse;
    const putDescriptor = getUpcomingEventsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(eventsFetched(response)));
  });

  it('should call the eventsError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getUpcomingEventsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(eventsFetchError(response)));
  });
});

/*
* Tests for pagination etc...
*/
