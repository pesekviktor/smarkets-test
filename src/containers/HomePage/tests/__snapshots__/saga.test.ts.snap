// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEvents Saga should call the eventsError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/v3/events/?state=upcoming&type=football",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getEvents Saga should dispatch the eventsFetched action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/v3/events/?state=upcoming&type=football",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
